integration/simulation/simulation_test.go

```diff
@@ -30,21 +34,62 @@ func TestSimulation(t *testing.T) {
 	defer f.Close()
 	common.SetLog(f)
 
-	blockDuration := uint64(20_000)
-	l1netw, l2netw := RunSimulation(5, 10, 15, blockDuration, blockDuration/15, blockDuration/3)
-	firstNode := l2netw.nodes[0]
-	checkBlockchainValidity(t, l1netw, l2netw, firstNode.Enclave.TestDB(), firstNode.Enclave.TestPeekHead().Head)
-	stats := l1netw.Stats
+	txGenerator := NewTransactionGenerator(5)
+
+	numberOfNodes := 10
+	simulationTime := 15
+	avgBlockDuration := uint64(20_000)
+	avgLatency := avgBlockDuration / 15
+	avgGossipPeriod := avgBlockDuration / 3
+
+	stats := NewStats(numberOfNodes, simulationTime, avgBlockDuration, avgLatency, avgGossipPeriod)
+
+	l1Network, l2Network := RunSimulation(txGenerator, numberOfNodes, simulationTime, avgBlockDuration, avgBlockDuration/15, avgBlockDuration/3, stats)
+
+	checkBlockchainValidity(t, l1Network, l2Network, txGenerator)
+
+	fmt.Println("Simulation ended...")
 	fmt.Printf("%+v\n", stats)
 	// pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
 }
 
-func checkBlockchainValidity(t *testing.T, l1Network L1NetworkCfg, l2Network L2NetworkCfg, db enclave2.DB, r *enclave2.Rollup) {
-	stats := l1Network.Stats
-	p := r.Proof(db)
-	validateL1(t, p, stats, db)
-	totalWithdrawn := validateL2(t, r, stats, db)
-	validateL2State(t, l2Network, stats, totalWithdrawn)
+func checkBlockchainValidity(t *testing.T, l1Network L1NetworkCfg, l2Network L2NetworkCfg, txManager *TransactionManager) {
+	// TODO check all nodes are the same height ?
+	// pick one node to draw height
+	enclaveNode := l2Network.nodes[0].Enclave
+	l1Node := l1Network.nodes[0]
+	l1Height := enclaveNode.L1Height()
+	l1HeightHash := enclaveNode.L1HeightHash()
+	l2Height := enclaveNode.L2Height()
+
+	validateL1(t, l1Network.Stats, l1Height, l1HeightHash, l1Node)
+	totalWithdrawn := validateL2Stats(t, l1Network.Stats, l2Height, txManager)
+	validateL2StateStats(t, l2Network, l1Network.Stats, totalWithdrawn, txManager.wallets)
+	validateL2State(t, l2Network.nodes, txManager)
+}
+
+// validateL2State tests that all transaction in the transaction Manager are found in the blockchain state of each node
+func validateL2State(t *testing.T, nodes []*obscuro_node.Node, txManager *TransactionManager) {
+	// Parallelize this check
+	var nGroup errgroup.Group
+
+	// Check the balance of all nodes adds up with the balance in the stats
```

Comment:
this comment is not describing what happens here

Reply:
Good catch !